### ISSUE TITLE
[STORM-650] Storm-Kafka Refactoring and Improvements

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#4491]
<p>This is intended to be a parent/umbrella JIRA covering a number of efforts/suggestions aimed at improving the storm-kafka module. The goal is to facilitate communication and collaboration by providing a central point for discussion and coordination.</p>

<p>The first phase should be to identify and agree upon a list of high-level points we would like to address. Once that is complete, we can move on to implementation/design discussions, followed by an implementation plan, division of labor, etc.</p>

<p>A non-exhaustive, initial list of items follows. New/additional thoughts can be proposed in the comments.</p>

<ul>
	<li>Improve API for Specifying the Kafka Starting point<br/>
Configuring the kafka spout's starting position (e.g. forceFromStart=true) is a common source of confusion. This should be refactored to provide an easy to understand, unambiguous API for configuring this property.</li>
</ul>

<ul>
	<li>Use Kafka APIs Instead of Internal ZK Metadata (<a href="https://issues.apache.org/jira/browse/STORM-590" title="KafkaSpout should use kafka consumer api" class="issue-link" data-issue-key="STORM-590"><del>STORM-590</del></a>)<br/>
Currently the Kafka spout relies on reading Kafka's internal metadata from zookeeper. This should be refactored to use the Kafka Consumer API to protect against changes to the internal metadata format stored in ZK.</li>
</ul>

<ul>
	<li>Improve Error Handling<br/>
There are a number of failure scenarios with the kafka spout that users may want to react to differently based on their use case. Add a failure handler API that allows users to implement and/or plug in alternative failure handling implementations. It is assumed that default (sane) implementations would be included and configured by default.</li>
</ul>

<ul>
	<li>Configuration/General Refactoring (BrokerHosts, etc.) (<a href="https://issues.apache.org/jira/browse/STORM-631" title="Simplifying kafka connector code" class="issue-link" data-issue-key="STORM-631"><del>STORM-631</del></a>)<br/>
(need to flesh this out better) Reduce unnecessary marker interfaces/"instance of" checks. Unify configuration of core storm/trident spout implementations.</li>
</ul>

<ul>
	<li>Kafka Spout doesn't pick up from the beginning of the queue unless forceFromStart specified (<a href="https://issues.apache.org/jira/browse/STORM-563" title="Kafka Spout doesn&#39;t pick up from the beginning of the queue unless forceFromStart specified" class="issue-link" data-issue-key="STORM-563"><del>STORM-563</del></a>)</li>
</ul>

<p>Discussion Items:</p>

<ul>
	<li>How important is backward compatibility?</li>
</ul>

---
<details><summary><i>Originally reported by <a title="P. Taylor Goetz" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz">ptgoetz</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-650" target="_blank">Storm-Kafka Refactoring and Improvements</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Hugo Da Cruz Louro" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro">hmclouro</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Abandoned
<li><b>resolved</b>: 2024-07-02T10:54:12+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ogorun">ogorun</a>:</i>
<p>Hi all,<br/>
Linking a variant of 'Improve Error Handling' implementation: <a href="https://github.com/apache/storm/pull/406" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/406</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=xavier.stevens">xavier.stevens</a>:</i>
<p>I think it would be useful if the kafka spout implementations supported topic whitelisting by regular expression topic matching. Kafka supports this in their high-level Consumer API.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=parth.brahmbhatt">parth.brahmbhatt</a>:</i>
<p>I don't think we are proposing to use high level consumer. Currently storm uses kafka's internal zk details to figure out a topic's partition -> List<Broker> and leader broker mapping. <a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a> is about changing that so we can use Kafka's admin API to get that information. The partition to consumer assignment is still handled by storm's own PartitionManager.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wurstmeister">wurstmeister</a>:</i>
<p>I’ve tried to put the current suggestions into a story format that should help us create/update the appropriate tickets.<br/>
I've grouped the suggestions into things that affect the internals and things that affect users of the library. </p>

<p>As a storm developer I’d like to use the Kafka Metadata API for broker management to avoid having to know kafka-internal zookeeper details (<a href="https://issues.apache.org/jira/browse/STORM-590" title="KafkaSpout should use kafka consumer api" class="issue-link" data-issue-key="STORM-590"><del>STORM-590</del></a>/<a href="https://issues.apache.org/jira/browse/STORM-631" title="Simplifying kafka connector code" class="issue-link" data-issue-key="STORM-631"><del>STORM-631</del></a>) <br/>
As a storm developer I’d like to use ITuple interface consistently to avoid duplication (<a href="https://issues.apache.org/jira/browse/STORM-631" title="Simplifying kafka connector code" class="issue-link" data-issue-key="STORM-631"><del>STORM-631</del></a>)<br/>
As a storm developer I’d like all loggers to be private (<a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>)<br/>
As a storm developer I’d like consistent, structured exception handling (<a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>)<br/>
As a storm developer I’d like to use the new kafka consumer API (0.8.3) to reduce dependencies and use long term supported kafka apis (<a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>)<br/>
As a storm developer I’d like to use the new kafka producer API to reduce dependencies and use long term supported kafka apis (<a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>)<br/>
As a storm developer I'd like to avoid having to use unnecessary marker interfaces (<a href="https://issues.apache.org/jira/browse/STORM-631" title="Simplifying kafka connector code" class="issue-link" data-issue-key="STORM-631"><del>STORM-631</del></a>) </p>


<p>As an API client developer I’d like to be able to distinguish between internal and public APIs to avoid confusion (<a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>)<br/>
As an API client developer I’d like to be able to select the starting point in kafka in an unambiguous way (<a href="https://issues.apache.org/jira/browse/STORM-563" title="Kafka Spout doesn&#39;t pick up from the beginning of the queue unless forceFromStart specified" class="issue-link" data-issue-key="STORM-563"><del>STORM-563</del></a>)<br/>
As an API client developer I’d like all public APIs to be documented (<a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>)<br/>
As an API client developer I’d like to be able to use a pluggable failure handler (<a href="https://github.com/apache/storm/pull/406" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/406</a>)<br/>
As an API client developer I’d like to be able to use a single way of configuring storm and trident kafka topologies (<a href="https://issues.apache.org/jira/browse/STORM-631" title="Simplifying kafka connector code" class="issue-link" data-issue-key="STORM-631"><del>STORM-631</del></a>)<br/>
As an API client developer I'd like the kafka related configuration to be immutable (<a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>)<br/>
As an API client developer I’d like to be able to white list topics in the kafka spout. (<a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>)<br/>
As an API client developer I’d like to know the offset and partition for a message so i can audit and replay messages (<a href="https://issues.apache.org/jira/browse/STORM-697" title="Support for Emitting Kafka Message Offset and Partition" class="issue-link" data-issue-key="STORM-697"><del>STORM-697</del></a>) </p>

<p>If we agree on this this list then I can go ahead and create/update the tickets so that we can move forward and start discussing further details in the individual tickets.</p>

<p>Thanks </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha">sriharsha</a>:</i>
<p>Thanks for summarizing <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wurstmeister" class="user-hover" rel="wurstmeister">Thomas Becker</a> .  There are lot of things here instead of getting this in as one patch can we do it progressively. <br/>
We've a PR up for <a href="https://issues.apache.org/jira/browse/STORM-631" title="Simplifying kafka connector code" class="issue-link" data-issue-key="STORM-631"><del>STORM-631</del></a> is it conflicting with where we want to go with kafka-connector If not we should get it in and open up additional jiras for further patches.<br/>
<a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=parth.brahmbhatt" class="user-hover" rel="parth.brahmbhatt">Parth Brahmbhatt</a> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz" class="user-hover" rel="ptgoetz">P. Taylor Goetz</a> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a> any thoughts . We are getting lot of queries around Kafka spout config IMO it definitely need refactoring I don't think this parent JIRA hold back on any progress.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz">ptgoetz</a>:</i>
<p>After a lot of thought and a romp through the kafka spout source code, I think we should consider an aggressive refactor/rewrite of the Kafka spout.</p>

<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wurstmeister" class="user-hover" rel="wurstmeister">Thomas Becker</a> has put together a pretty good list in agile story format that I think should be considered as a starting point.</p>

<p>The bottom line is that IMHO, first class Kafka support is pretty critical, and I don't think we're there. If users are asking the same questions over and over, something needs to be fixed either in code or documentation. I think it's both.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz">ptgoetz</a>:</i>
<p>Thanks <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jkreps" class="user-hover" rel="jkreps">Jay Kreps</a>, I think that sounds like a great idea. Can you point to some docs for the new consumer API?</p>

<p>I'll try to review the API and our code to see what, if anything, we would need.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz">ptgoetz</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jkreps" class="user-hover" rel="jkreps">Jay Kreps</a> nm... Found the docs you referenced earlier.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jkreps">jkreps</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz" class="user-hover" rel="ptgoetz">P. Taylor Goetz</a> Awesome, much appreciated.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wurstmeister">wurstmeister</a>:</i>
<p>Hi Jay, </p>

<p>could you please point me to some documentation that describes how the KafkaConsumer handles a leader change?<br/>
Is there any difference in the behaviour depending on whether I subscribe to a whole topic or individual TopicPartitions? </p>

<p>Thanks </p>

<p>Thomas </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jkreps">jkreps</a>:</i>
<p>Leadership handoff is totally transparent from the point of view of that client. All failover is done within the client. When a leader fails you might see a slight blip in latency but the client handles the process of discovering the new leader and fetching data from it. this is true irrespective of whether you subscribe to individual partitions or to the whole topic.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wurstmeister">wurstmeister</a>:</i>
<p>Thanks for the confirmation <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jkreps" class="user-hover" rel="jkreps">Jay Kreps</a>. While evaluating the new consumer for storm I came across a small issue with the leadership handoff when subscribing to individual partitions. I’ve put my test case on <a href="https://github.com/wurstmeister/kafka-client-test/blob/master/README.md" class="external-link" target="_blank" rel="nofollow noopener">github</a>. It would be great if you could let me know if <a href="https://github.com/wurstmeister/kafka-client-test/blob/master/README.md#test-case-2-per-partition-subscription" class="external-link" target="_blank" rel="nofollow noopener">"test case 2"</a> is a valid test case or if I have misunderstood the documentation (<a href="http://kafka.apache.org/083/javadoc/index.html?org/apache/kafka/clients/consumer/KafkaConsumer.html" class="external-link" target="_blank" rel="nofollow noopener">http://kafka.apache.org/083/javadoc/index.html?org/apache/kafka/clients/consumer/KafkaConsumer.html</a>). </p>

<p>In terms of compatibility it looks like the new client is compatible with kafka < 0.8.3 if we only subscribe to TopicPartitions because that does not use the new join API. Is that correct?</p>

<p>Thanks</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jkreps">jkreps</a>:</i>
<p>Hey <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wurstmeister" class="user-hover" rel="wurstmeister">Thomas Becker</a> yeah that test case will be covered. That client is still under very active development and the system level testing isn't started yet so I wouldn't worry too much about any current deficiencies aside from API  shortcomings.</p>

<p>You are correct that subscribing to complete topics requires the server-side partition balancing feature that will be in 0.8.3. Partition-level consumption just depends on the fetch and metadata requests which are unchanged in the 0.8.x releases. I believe, but we would have to check that offset commit is there from 0.8.1 on if you guys would be using the offset commit apis.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wurstmeister">wurstmeister</a>:</i>
<p>Thanks <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jkreps" class="user-hover" rel="jkreps">Jay Kreps</a>. <br/>
As far as I can see the new consumer api offers everything we need. Especially the transparent leader lookup/handoff will simplify things.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jkreps">jkreps</a>:</i>
<p>Great, really appreciate your looking at that!</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/406#issuecomment-111192307" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/406#issuecomment-111192307</a></p>

<p>    > Is it better to open separate JIRA issue or this pull request is considered as part of subjects discussed in <a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>?</p>

<p>    @ogorun  I am sorry I did not see the notification for you comment earlier.  Yes, I think the place to discuss this is <a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>.  We might ask on <a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a> whether we should add a new JIRA Issue for this to the <a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a> epic.  We might ask on <a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a> whether we should add a new JIRA Issue for this to the <a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a> epic.  So far, I am not sure this specific issue is addressed in <a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>, and it seems some kafka API changes are coming as well.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sunh11373">sunh11373</a>:</i>
<p>Hi all,</p>

<p>I submitted   <a href="https://github.com/apache/storm/pull/703" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/703</a></p>

<p>The PR is about storing spout state, including offsets, inside Kafka instead of ZK using Kafka's offset management api.  Not sure whether it is also in the scope of Use Kafka APIs Instead of Internal ZK Metadata (<a href="https://issues.apache.org/jira/browse/STORM-590" title="KafkaSpout should use kafka consumer api" class="issue-link" data-issue-key="STORM-590"><del>STORM-590</del></a>).  I added it mainly to make it work with existing offset monitoring tools.  The change is completely backward compatible.</p>

<p>Thanks</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/703#issuecomment-135318226" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/703#issuecomment-135318226</a></p>

<p>    @hsun-cnnxty <br/>
    Since <a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a> is a kind of EPIC / Umbrella issue, and it seems not maintained now, so it would be better to file new issue.<br/>
    And please do not modify version from pom.xml.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/406#issuecomment-167632103" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/406#issuecomment-167632103</a></p>

<p>    @ogorun, how should we move forward with this work?  Has this issue been resolved as part of <a href="https://issues.apache.org/jira/browse/STORM-650" title="Storm-Kafka Refactoring and Improvements" class="issue-link" data-issue-key="STORM-650"><del>STORM-650</del></a>, or does it still need to be addressed?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rzo1">rzo1</a>:</i>
<p>We are in the process of migrating our issue tracking system to GitHub issues. This issue has been stale for the last 5 years, and STORM has evolved with many dependency updates since this issue was last updated.</p>

<p>If the issue persists in the latest version of Storm, please feel free to reopen it or provide new context. Thank you very much!</p>
